atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/asts/DefaultSingleBucketSweepTask.java

```diff
@@ -0,0 +1,134 @@
+import java.util.function.LongSupplier;
+
+public class DefaultSingleBucketSweepTask implements SingleBucketSweepTask {
+    private static final SafeLogger log = SafeLoggerFactory.get(DefaultSingleBucketSweepTask.class);
+
+    private final BucketProgressStore bucketProgressStore;
+    private final SweepQueueReader sweepQueueReader;
+    private final SweepQueueDeleter sweepQueueDeleter;
+    private final SweepQueueCleaner sweepQueueCleaner;
+    private final LongSupplier sweepTimestampSupplier;
+    private final TargetedSweepMetrics targetedSweepMetrics;
+
+    public DefaultSingleBucketSweepTask(
+            BucketProgressStore bucketProgressStore,
+            SweepQueueReader sweepQueueReader,
+            SweepQueueDeleter sweepQueueDeleter,
+            SweepQueueCleaner sweepQueueCleaner,
+            LongSupplier sweepTimestampSupplier,
+            TargetedSweepMetrics targetedSweepMetrics) {
+        this.bucketProgressStore = bucketProgressStore;
+        this.sweepQueueReader = sweepQueueReader;
+        this.sweepQueueDeleter = sweepQueueDeleter;
+        this.sweepQueueCleaner = sweepQueueCleaner;
+        this.sweepTimestampSupplier = sweepTimestampSupplier;
+        this.targetedSweepMetrics = targetedSweepMetrics;
+    }
+
+    @Override
+    public long runOneIteration(SweepableBucket sweepableBucket) {
+        long sweepTimestampForIteration = sweepTimestampSupplier.getAsLong();
+        long minTimestampForPartition = sweepableBucket.bucketIdentifier() * SweepQueueUtils.TS_FINE_GRANULARITY;
+        if (sweepTimestampForIteration <= minTimestampForPartition) {
+            // This means that the sweep timestamp has not entered this partition yet, so we do not need to process
+            // anything. Note that sweep timestamps are exclusive, so <= is correct.
+            return 0L;
+        }
+
+        BucketProgress existingBucketProgress =
+                bucketProgressStore.getBucketProgress(sweepableBucket).orElse(BucketProgress.INITIAL_PROGRESS);
+
+        if (existingBucketProgress.isBucketCompletelySwept()) {
+            // The bucket has been fully swept. It might still be returned here if we accidentally thought it was a
+            // candidate. In any case, there is no work to be done here: this bucket is done.
+            return 0L;
+        }
+
+        long lastSweptTimestampInPartition = minTimestampForPartition + existingBucketProgress.timestampOffset();
+        if (sweepTimestampForIteration <= lastSweptTimestampInPartition) {
+            // This is a bit unusual but can happen if we have some mechanism of initialising bucket progress beyond
+            // zero.
+            return 0L;
+        }
+
+        long partitionEndExclusive = Math.min(
+                sweepTimestampForIteration,
+                SweepQueueUtils.minTsForFinePartition(sweepableBucket.bucketIdentifier() + 1));
+
+        // TODO (jkong): Don't just use this, it does a lot of unnecessary stuff
```

Comment:
The intention here was because buckets used to be perfectly aligned on partition boundaries. This is no longer the case, and so this is necessary